### PR TITLE
Trying to fix dependency shutdown

### PR DESCRIPTION
**Pull request by SamGondelman**
https://highfidelity.manuscript.com/f/cases/16254/DependencyManager-should-validate-that-a-dependency-is-actually-destroyed-when-it-s-asked

Test plan:
- Verify that the tablet and audio still work.  Verify that you can run and stop scripts.
- Verify that you don't crash on shutdown.  After shutdown, check your logs.  Near the end, you should see "DependencyManager::destroy(): class ScriptEngines was not properly destroyed!" but you shouldn't see any other warnings about `DependencyManager::destroy()`.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/11152/HighFidelity-Beta-PR13555-ec59108.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/11152/HighFidelity-Beta-PR13555-ec59108.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/11152/HighFidelity-Beta-Interface-PR13555-ec59108.dmg

File: libraries/script-engine/src/ScriptEngines.cpp [hunk 1]
```diff
@@ -564,10 +563,10 @@ int ScriptEngines::runScriptInitializers(ScriptEnginePointer scriptEngine) {
 
 void ScriptEngines::launchScriptEngine(ScriptEnginePointer scriptEngine) {
     connect(scriptEngine.data(), &ScriptEngine::finished, this, &ScriptEngines::onScriptFinished, Qt::DirectConnection);
-    connect(scriptEngine.data(), &ScriptEngine::loadScript, [&](const QString& scriptName, bool userLoaded) {
+    connect(scriptEngine.data(), &ScriptEngine::loadScript, [this](const QString& scriptName, bool userLoaded) {
```
Comment (NissimHadar) [inline libraries/script-engine/src/ScriptEngines.cpp]: Why did you change this?  Is the & wrong?

Comment (SamGondelman) [reply libraries/script-engine/src/ScriptEngines.cpp]: This one doesn't matter as much.  `[&]` will capture scriptEngine by reference, whereas `[this]` won't.  Capturing the engine by reference should have been fine because it wouldn't increase the ref count, but there's also no harm in removing the unintentional capture.

Comment (SamGondelman) [reply libraries/script-engine/src/ScriptEngines.cpp]: Correct me if I'm wrong, but I think `&` will effectively capture this *and* scriptEngine, whereas `this` just captures `this` (which is what the lamda requires).

Comment (NissimHadar) [reply libraries/script-engine/src/ScriptEngines.cpp]: Exactly - so you get more access.  I though you were trying to minimize the access.

Comment (SamGondelman) [reply libraries/script-engine/src/ScriptEngines.cpp]: [&] captures `this` *and* `scriptEngine`, `[this]` just captures `this`, so we've removed a capture of scriptEngine.

Comment (NissimHadar) [reply libraries/script-engine/src/ScriptEngines.cpp]: Oh, now I see your point.

Comment (Atlante45) [reply libraries/script-engine/src/ScriptEngines.cpp]: @SamGondelman `&` and `=` only capture what's used, so even if `=` had been used, it wouldn't bump the ref count unless the variable is used in the lambda.
(Just wanted to clarify that, I support this changes, it restrains the lambda's scope, which is always safer)

Comment (SamGondelman) [reply libraries/script-engine/src/ScriptEngines.cpp]: Ah, ok!  Good to know.  Most of these changes have absolutely no effect then.

File: libraries/shared/src/DependencyManager.h [hunk 2]
```diff
@@ -139,7 +139,13 @@ QSharedPointer<T> DependencyManager::set(Args&&... args) {
 template <typename T>
 void DependencyManager::destroy() {
     static size_t hashCode = manager().getHashCode<T>();
-    manager().safeGet(hashCode).clear();
+    QSharedPointer<Dependency>& shared = manager().safeGet(hashCode);
+    QWeakPointer<Dependency> weak = shared;
+    shared.clear();
+    // Check that the dependency was actually destroyed.  If it wasn't, it was improperly captured somewhere
+    if (weak.lock()) {
+        qWarning() << "DependencyManager::destroy(): Dependency was not properly destroyed!";
```
Comment (Atlante45) [inline libraries/shared/src/DependencyManager.h]: You should print the typeid name like in `DependencyManager::get` to help during  debug.

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/11220/HighFidelity-Beta-PR13555-d2699e8.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/11220/HighFidelity-Beta-PR13555-d2699e8.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/11220/HighFidelity-Beta-Interface-PR13555-d2699e8.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/11220/HighFidelity-Beta-PR13555-d2699e8.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/11220/HighFidelity-Beta-Interface-PR13555-d2699e8.exe

**Comment by qlimnavarro** [comment]
Currently Testing

**Comment by qlimnavarro** [comment]
Fail @conklin94122 
Mac still crashes when closing.
[hifi-log_2018-07-11_15.59.23.txt](https://github.com/highfidelity/hifi/files/2186379/hifi-log_2018-07-11_15.59.23.txt)
[hifi-log_2018-07-11_15.58.09.txt](https://github.com/highfidelity/hifi/files/2186380/hifi-log_2018-07-11_15.58.09.txt)
[hifi-log_2018-07-11_16.00.10.txt](https://github.com/highfidelity/hifi/files/2186381/hifi-log_2018-07-11_16.00.10.txt)

**Comment by SamGondelman** [comment]
@qlimnavarro Hmmmm this PR doesn't actually explicitly fix any crashes, only potential ones.  Do you see more crashes with this PR compared to master?  If it's the same or fewer, I would still call this a pass.  It's weird that I don't see the destroy warning in your logs though...maybe we should check it on backtrace.

**Comment by Qliemillar** [comment]
Build this please

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/11363/HighFidelity-Beta-PR13555-d2699e8.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/11363/HighFidelity-Beta-PR13555-d2699e8.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/11363/HighFidelity-Beta-Interface-PR13555-d2699e8.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/11363/HighFidelity-Beta-PR13555-d2699e8.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/11363/HighFidelity-Beta-Interface-PR13555-d2699e8.exe

**Comment by qliacoombs** [comment]
Build this please

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/11489/HighFidelity-Beta-PR13555-d2699e8.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/11489/HighFidelity-Beta-PR13555-d2699e8.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/11489/HighFidelity-Beta-Interface-PR13555-d2699e8.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/11489/HighFidelity-Beta-PR13555-d2699e8.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/11489/HighFidelity-Beta-Interface-PR13555-d2699e8.exe

**Comment by conklin94122** [comment]
build this please @conklin94122

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/11577/HighFidelity-Beta-PR13555-d2699e8.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/11577/HighFidelity-Beta-PR13555-d2699e8.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/11577/HighFidelity-Beta-Interface-PR13555-d2699e8.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/11577/HighFidelity-Beta-PR13555-d2699e8.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/11577/HighFidelity-Beta-Interface-PR13555-d2699e8.exe

**Comment by SamGondelman** [comment]
build this please

**Comment by SamGondelman** [comment]
@conklin94122 can I have someone retest this on Mac and confirm that it doesn't increase the rate of crashes on exit?

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/11725/HighFidelity-Beta-PR13555-d2699e8.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/11725/HighFidelity-Beta-PR13555-d2699e8.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/11725/HighFidelity-Beta-Interface-PR13555-d2699e8.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/11725/HighFidelity-Beta-PR13555-d2699e8.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/11725/HighFidelity-Beta-Interface-PR13555-d2699e8.exe

**Comment by qaAdamSmith82** [comment]
Fail @conklin94122 

Mac does crash more.  I open/closed the Mac and PC clients ~20 times.  The PC build crashed about 50% of the time.  The Mac build crashed every time but one.

The manner in which they crashed was different as well.  For lack of a better term, the PC build crashed gracefully, the only indication that a crash had occurred was the toast on startup asking the user if they want to reset settings because they crashed last run.  The Mac hangs, locks and then the user is given the "interface.exe has unexpectedly stopped" message on every exit  

Attached files:
1. hifi-log_PC13555
2. hifi-log_MAC13555

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/11803/HighFidelity-Beta-PR13555-1ba28de.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/11803/HighFidelity-Beta-PR13555-1ba28de.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/11803/HighFidelity-Beta-Interface-PR13555-1ba28de.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/11803/HighFidelity-Beta-PR13555-1ba28de.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/11803/HighFidelity-Beta-Interface-PR13555-1ba28de.exe

**Comment by qaAdamSmith82** [comment]
Build this please

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/11817/HighFidelity-Beta-PR13555-1ba28de.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/11817/HighFidelity-Beta-PR13555-1ba28de.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/11817/HighFidelity-Beta-Interface-PR13555-1ba28de.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/11817/HighFidelity-Beta-PR13555-1ba28de.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/11817/HighFidelity-Beta-Interface-PR13555-1ba28de.exe

**Comment by qaAdamSmith82** [comment]
build this please @conklin94122

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/11824/HighFidelity-Beta-PR13555-1ba28de.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/11824/HighFidelity-Beta-Interface-PR13555-1ba28de.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/11824/HighFidelity-Beta-PR13555-1ba28de.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/11824/HighFidelity-Beta-Interface-PR13555-1ba28de.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/11826/HighFidelity-Beta-PR13555-3c7f1a0.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/11826/HighFidelity-Beta-PR13555-3c7f1a0.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/11826/HighFidelity-Beta-Interface-PR13555-3c7f1a0.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/11826/HighFidelity-Beta-PR13555-3c7f1a0.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/11826/HighFidelity-Beta-Interface-PR13555-3c7f1a0.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/11871/HighFidelity-Beta-PR13555-2edefe3.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/11871/HighFidelity-Beta-PR13555-2edefe3.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/11871/HighFidelity-Beta-Interface-PR13555-2edefe3.exe

**Comment by qaAdamSmith82** [comment]
Fail @conklin94122 

Crash/Mac

If the user attempts to enter a domain, and gets either a protocol mismatch, or is unable to enter for any reason (not logged in, etc), clicking the interface menu in the upper left causes a crash.

5/5 repro

Attached files:
1.Mac_13555_1.txt
[Mac_13555_1.txt](https://github.com/highfidelity/hifi/files/2263607/Mac_13555_1.txt)
[Mac_13555_2.txt](https://github.com/highfidelity/hifi/files/2263608/Mac_13555_2.txt)


2.Mac_13555_2.txt

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/11879/HighFidelity-Beta-PR13555-e3756a6.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/11879/HighFidelity-Beta-PR13555-e3756a6.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/11879/HighFidelity-Beta-Interface-PR13555-e3756a6.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/11879/HighFidelity-Beta-PR13555-e3756a6.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/11879/HighFidelity-Beta-Interface-PR13555-e3756a6.exe

**Comment by conklin94122** [comment]
build this please

**Comment by conklin94122** [comment]
QA Approval is stale, retesting.

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12143/HighFidelity-Beta-PR13555-e3756a6.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12143/HighFidelity-Beta-PR13555-e3756a6.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12143/HighFidelity-Beta-Interface-PR13555-e3756a6.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12143/HighFidelity-Beta-PR13555-e3756a6.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12143/HighFidelity-Beta-Interface-PR13555-e3756a6.dmg

**Comment by qlimnavarro** [comment]
Currently Testing

**Comment by qlimnavarro** [comment]
Pass @conklin94122 
Mac does not mention DependencyManager other than its intended purpose. And does not crash.
